src/hotspot/share/services/memReporter.hpp

```diff
@@ -66,11 +66,22 @@ class MemReporterBase : public StackObj {
   }
 
   // Convert diff amount in bytes to current reporting scale
-  inline long diff_in_current_scale(size_t s1, size_t s2) const {
-    long amount = (long)(s1 - s2);
-    long scale = (long)_scale;
-    amount = (amount > 0) ? (amount + scale / 2) : (amount - scale / 2);
-    return amount / scale;
+  inline int64_t diff_in_current_scale(size_t s1, size_t s2) const {
+    // _scale should not be 0, otherwise division by zero at return.
+    assert(_scale != 0, "wrong scale");
+
+    bool is_negative = false;
+    if (s1 < s2) {
+      is_negative = true;
+      swap(s1, s2);
+    }
+
+    size_t amount = s1 - s2;
+    assert(amount <= SIZE_MAX - _scale / 2, "size_t overflow");
```

Comment:
This assert can fail for sufficiently large `amount`, even though the desired
value is computable.  The value we want is
```
size_t scaled = (amount + _scale/2) / _scale;
```
rounded to nearest, without risk of overflow. We can split `amount` into `p + q`,
where `q = amount % _scale` and `p = amount - q` (which is also `(amount / _scale) * _scale`).
Then use
```
size_t scaled = (p + q + _scale/2) / _scale;
```
=>
```
size_t scaled = (p / _scale) + ((q + _scale/2) / _scale);
```
The lefthand side of the addition is exact.
The righthand side is 0 if `q <= (_scale - 1)/2`, else 1. (The -1 is to
account for odd _scale values.)  So
```
size_t scaled = (amount / _scale);
if ((amount % _scale) <= (_scale - 1)/2) {
  scaled += 1;
}
```

Reply:
And after all that I wrote the conditional backward.  That should be `(amount % _scale) > (_scale - 1)/2`.

Reply:
Save future readers the trouble of reverse engineering how this works.  Add a comment with some info
about the derivation or something.